src/transform/src/projection_pushdown.rs

```diff
@@ -0,0 +1,401 @@
+                };
+                reverse_permute_columns(
+                    group_key
+                        .iter_mut()
+                        .chain(order_key.iter_mut().map(|o| &mut o.column)),
+                    columns_to_pushdown.iter(),
+                );
+                self.action(input, columns_to_pushdown.clone(), gets);
+                columns_to_pushdown
+            }
+            MirRelationExpr::Negate { input } => {
+                self.action(input, desired_projection.clone(), gets);
+                desired_projection.clone()
+            }
+            MirRelationExpr::Union { base, inputs } => {
+                self.action(base, desired_projection.clone(), gets);
+                for input in inputs {
+                    self.action(input, desired_projection.clone(), gets);
+                }
+                desired_projection.clone()
+            }
+            MirRelationExpr::Threshold { input } => {
+                // Threshold requires all columns, as collapsing any distinct values
+                // has the potential to change how it thresholds counts. This could
+                // be improved with reasoning about distinctness or non-negativity.
+                let arity = input.arity();
+                self.action(input, (0..arity).collect(), gets);
+                (0..arity).collect()
+            }
+            MirRelationExpr::ArrangeBy { input, keys } => {
+                // If a key set contains keys that are not demanded upstream,
+                // delete the key set.
+                // If `keys` becomes empty, delete the arrangeby, and keep
+                // pushing the projection. Otherwise, do not push the project past
+                // the ArrangeBy
+                for i in (0..keys.len()).rev() {
+                    if keys[i]
+                        .iter()
+                        .any(|k| k.support().iter().any(|c| !desired_projection.contains(c)))
+                    {
+                        keys.remove(i);
+                    }
+                }
+                if keys.is_empty() {
+                    self.action(input, desired_projection.clone(), gets);
+                    *relation = input.take_dangerous();
+                    desired_projection.clone()
+                } else {
+                    let arity = input.arity();
+                    self.action(input, (0..arity).collect(), gets);
+                    (0..arity).collect()
+                }
+            }
+            MirRelationExpr::DeclareKeys { input, keys } => {
+                // TODO[btv] - If and when we add a "debug mode" that asserts whether this is truly a key,
+                // we will probably need to add the key to the set of demanded
+                // columns.
+
+                // Current behavior is that if a key is not contained with the
+                // desired_projection, then it is not relevant to the query plan
+                // and can be removed.
```

Comment:
This is a bit subtle, but might check out. For example, reduce elision wants to know about keys because then it can remove `Reduce` instances. The logic doesn't require that the reduce key is exactly a unique key, but that it contains one. That does mean it has to reference each column though, so it should be ok. Are there other cases where we might not reference a key, but still rely on one existing, or where projecting away a column that makes for a unique key leaves us with fewer options? If we leave in columns that are part of marked keys, rather than removing the keys, does that make plans bad (or the other way around?).

Reply:
Created issue #8159 so we remember to holistically investigate how the optimizer should be handling `MirRelationExpr::DeclareKeys`. 

For this PR, I'm just keeping the original behavior from `Demand` transform. Since the keys of `DeclareKeys` are not being added to the `demand`, I'm not adding the keys to `desired_projection`. If I'm not adding the keys to `desired_projection`, I have to delete them from `DeclareKeys` because otherwise they would be pointing to nonexistent columns.

Reply:
Works for me!